src/lib/sql/sql_pipeline_statement.cpp

```diff
@@ -126,41 +173,64 @@ const std::shared_ptr<SQLQueryPlan>& SQLPipelineStatement::get_query_plan() {
+      // Handle query plan if statement has been cached
+      auto& plan = *cached_plan;
+
+      DebugAssert(!plan.tree_roots().empty(), "QueryPlan retrieved from cache is empty.");
+      if (plan.tree_roots().front()->transaction_context_is_set()) {
+        Assert(_use_mvcc == UseMvcc::Yes, "Trying to use MVCC cached query without a transaction context.");
+      } else {
+        Assert(_use_mvcc == UseMvcc::No, "Trying to use non-MVCC cached query with a transaction context.");
+      }
+
+      _query_plan->append_plan(plan.recreate());
+    } else if (const auto* execute_statement = dynamic_cast<const hsql::ExecuteStatement*>(statement)) {
+      // Handle query plan if we are executing a prepared statement
+      Assert(_prepared_statements, "Cannot execute statement without prepared statement cache.");
+      const auto plan = _prepared_statements->try_get(execute_statement->name);
+      Assert(plan, "Requested prepared statement does not exist!");
+
+      // Get list of arguments from EXECUTE statement.
+      std::vector<AllParameterVariant> arguments;
+      if (execute_statement->parameters != nullptr) {
+        for (const auto* expr : *execute_statement->parameters) {
+          arguments.push_back(HSQLExprTranslator::to_all_parameter_variant(*expr));
+        }
+      }
+
+      Assert(arguments.size() == plan->num_parameters(),
+             "Number of arguments provided does not match expected number of arguments.");
+
+      _query_plan->append_plan(plan->recreate(arguments));
     } else {
-      Assert(_use_mvcc == UseMvcc::No, "Trying to use non-MVCC cached query with a transaction context.");
-    }
+      // "Normal" mode in which the query plan is created
+      const auto& lqp = get_optimized_logical_plan();
+      _query_plan->add_tree_by_root(LQPTranslator{}.translate_node(lqp));
 
-    _query_plan->append_plan(plan.recreate());
-    if (_use_mvcc == UseMvcc::Yes) _query_plan->set_transaction_context(_transaction_context);
-
-    const auto done = std::chrono::high_resolution_clock::now();
-    _compile_time_micros = std::chrono::duration_cast<std::chrono::microseconds>(done - started);
-
-    return _query_plan;
-  }
-
-  const auto& lqp = get_optimized_logical_plan();
+      // Set number of parameters to match later in case of prepared statement
+      _query_plan->set_num_parameters(_num_parameters);
+    }
 
-  try {
-    _query_plan->add_tree_by_root(LQPTranslator{}.translate_node(lqp));
     if (_use_mvcc == UseMvcc::Yes) _query_plan->set_transaction_context(_transaction_context);
   } catch (const std::exception& exception) {
     throw std::runtime_error("Error while translating query plan:\n  " + std::string(exception.what()));
   }
 
-  const auto done = std::chrono::high_resolution_clock::now();
-  _compile_time_micros = std::chrono::duration_cast<std::chrono::microseconds>(done - started);
+  if (const auto* prepared_statement = dynamic_cast<const hsql::PrepareStatement*>(statement)) {
```

Comment:
The nesting in here is confusing. Why are some parts in the try-block and others not?

Reply:
Any suggestions on how to change this? The try catch block contains the 3 ways to get a QueryPlan and all of those can fail in some way that we want to catch. The rest is not relevant for the try-catch... Not sure how to change this to make it more intuitive. The only thing is can come up with is separate methods

Reply:
Mmh, ok. Let's leave it for now.